src/components/shared/thread/SampleGraph.js

```diff
@@ -185,35 +212,100 @@ export class ThreadSampleGraphImpl extends PureComponent<Props> {
     drawSamples(idleSamples, lighterBlue);
   }
 
-  _onClick = (event: SyntheticMouseEvent<>) => {
-    const canvas = this._canvas;
-    if (canvas) {
-      const { rangeStart, rangeEnd, thread } = this.props;
-      const r = canvas.getBoundingClientRect();
-
-      const x = event.pageX - r.left;
-      const time = rangeStart + (x / r.width) * (rangeEnd - rangeStart);
-
-      const sampleIndex = getSampleIndexClosestToCenteredTime(
-        thread.samples,
-        time
-      );
-
-      if (thread.samples.stack[sampleIndex] === null) {
-        // If the sample index refers to a null sample, that sample
-        // has been filtered out and means that there was no stack bar
-        // drawn at the place where the user clicked. Do nothing here.
-        return;
-      }
+  _onClick = (event: SyntheticMouseEvent<HTMLCanvasElement>) => {
+    const hoveredSample = this._getSampleAtMouseEvent(event);
+    this.props.onSampleClick(event, hoveredSample?.sample ?? null);
+  };
 
-      this.props.onSampleClick(event, sampleIndex);
+  _onMouseLeave = () => {
+    this.setState({ hoveredPixelState: null });
+  };
+
+  _onMouseMove = (event: SyntheticMouseEvent<HTMLCanvasElement>) => {
+    const canvas = event.currentTarget;
+    if (!canvas) {
+      return;
     }
+
+    const rect = canvas.getBoundingClientRect();
+    this.setState({
+      hoveredPixelState: this._getSampleAtMouseEvent(event),
+      mouseX: event.pageX,
+      // Have the tooltip align to the bottom of the track.
+      mouseY: rect.bottom - MOUSE_OFFSET,
+    });
   };
 
+  _getSampleAtMouseEvent(
+    event: SyntheticMouseEvent<HTMLCanvasElement>
+  ): null | HoveredPixelState {
+    const canvas = this._canvas;
+    if (!canvas) {
+      return null;
+    }
+
+    const { rangeStart, rangeEnd, thread, interval } = this.props;
+    const r = canvas.getBoundingClientRect();
+
+    const x = event.pageX - r.left;
```

Comment:
optional: `event.offsetX` is equal to this value. But I don't remember if it's available on the React's event or if we need to access it with `nativeEvent.offsetX`. (I think this was the case before but it may be available now)

(optional because we need to call getBoundingClientRect to get the width anyway, so it's only a matter of code style)